File: docs/snippets/ov_hetero.cpp
#include <openvino/runtime/core.hpp>
#include <openvino/core/model.hpp>


int main() {
ov::Core core;
auto model = core.read_model("sample.xml");
//! [part0]
for (auto && op : model->get_ops()) {
    op->get_rt_info()["affinity"] = "CPU";
}
//! [part0]

//! [part1]
// This example demonstrates how to perform default affinity initialization and then
// correct affinity manually for some layers
const std::string device = "HETERO:GPU,CPU";

// query_model result contains maping of supported operations to device
auto supported = core.query_model(model, device);

// update default affinities
supported["layerName"] = "CPU";

// set affinities to network
for (auto&& node : model->get_ops()) {
    auto& affinity = supported[node->get_friendly_name()];
    // Store affinity mapping using node runtime information
    node->get_rt_info()["affinity"] = affinity;
}

// load network with affinities set before
auto compiled_model = core.compile_model(model, device);
//! [part1]

//! [part2]
{
    auto compiled_model = core.compile_model(model, "HETERO:GPU,CPU");
}
{
    auto compiled_model = core.compile_model(model, "HETERO", ov::device::priorities("GPU", "CPU"));
}
{
    auto compiled_model = core.compile_model(model, "HETERO", ov::device::priorities("GPU,CPU"));
}
//! [part2]
return 0;
}

File: docs/snippets/ov_hetero.py
import openvino.runtime as ov

core = ov.Core()
model = core.read_model("sample.xml")

#! [set_manual_affinities]
for op in model.get_ops():
    rt_info = op.get_rt_info()
    rt_info["affinity"] = "CPU"
#! [set_manual_affinities]

#! [fix_automatic_affinities]
# This example demonstrates how to perform default affinity initialization and then
# correct affinity manually for some layers
device = "HETERO:GPU,CPU"

# query_model result contains mapping of supported operations to devices
supported_ops = core.query_model(model, device)

# update default affinities manually for specific operations
supported_ops["operation_name"] = "CPU"

# set affinities to a model
for node in model.get_ops():
    affinity = supported_ops[node.get_friendly_name()]
    node.get_rt_info()["affinity"] = "CPU"

# load model with manually set affinities
compiled_model = core.compile_model(model, device)
#! [fix_automatic_affinities]

#! [compile_model]
compiled_model = core.compile_model(model, device_name="HETERO:GPU,CPU")
# device priorities via configuration property
compiled_model = core.compile_model(model, device_name="HETERO", config={"MULTI_DEVICE_PRIORITIES": "GPU,CPU"})
#! [compile_model]

#! [configure_fallback_devices]
core.set_property("HETERO", {"MULTI_DEVICE_PRIORITIES": "GPU,CPU"})
core.set_property("GPU", {"PERF_COUNT": "YES"})
core.set_property("CPU", {"INFERENCE_PRECISION_HINT": "f32"})
compiled_model = core.compile_model(model=model, device_name="HETERO")
#! [configure_fallback_devices]
